contracts/proposal/dao-proposal-condorcet/src/m.rs

```diff
@@ -0,0 +1,382 @@
+    //
+    // which has a square root if you try and extract n from it.
+    pub n: u32,
+}
+
+pub(crate) enum Stats {
+    PositiveColumn {
+        /// Index of the column that is positive.
+        col: u32,
+        /// Smallest value in the column's distance from zero.
+        min_margin: Uint128,
+    },
+    NoPositiveColumn {
+        /// False if there exists a column where:
+        ///
+        /// distance_from_positivity(col) <= power_outstanding * (N-1)
+        /// && max_negative_magnitude(col) < power_outstanding
+        no_winnable_columns: bool,
+    },
+}
+
+impl M {
+    pub fn new(n: u32) -> Self {
+        M {
+            // example 4x4 M:
+            //
+            //  \  1  2  3
+            // -1  \  4  5
+            // -2 -4  \  6
+            // -3 -5 -6  \
+            //
+            // `cells` stores all the values in the upper diagonal of
+            // M. there are `N-1 + N-2 .. + 1` or `N (N-1) / 2` cells.
+            cells: vec![Cell::default(); (n * (n - 1) / 2) as usize],
+            n,
+        }
+    }
+
+    /// Gets the index in `self.cells` that corresponds to the index (x, y) in M.
+    ///
+    /// Invariant: x > y as otherwise the upper diagonal
+    /// (`self.cells`) will not contain an entry for the index.
+    fn index(&self, (x, y): (u32, u32)) -> u32 {
+        let n = self.n;
+        // the start of the row in `self.cells`.
+        //
+        // the easiest way to conceptualize this is
+        // geometrically. `y*n` is the area of the whole matrix up to
+        // row `y`, and thus the start index of the row if
+        // `self.cells` was not diagonalized [1]. `(y + 1) * y / 2` is the
+        // area of the space that is not in the upper diagonal.
+        //
+        // whole_area - area_of_non_diagonal = area_of_diagonal
+        //
+        // because we're in the land of discrete math and things are
+        // zero-indexed, area_of_diagonal == start_of_row.
+        let row = y * n - (y + 1) * y / 2;
+        // we know that x > y, so to get the index in `self.cells` we
+        // offset x by the distance of x from the line x = y (the
+        // diagonal), as `self.cells`' first index corresponds to the
+        // first item in that row of the upper diagonal.
+        let offset = x - (y + 1);
+        row + offset
+    }
+
+    pub(crate) fn get(&self, (x, y): (u32, u32)) -> Cell {
+        if x < y {
+            self.get((y, x)).invert()
+        } else {
+            let i = self.index((x, y)) as usize;
+            self.cells[i]
+        }
+    }
+
+    pub fn increment(&mut self, (x, y): (u32, u32), amount: Uint128) {
+        debug_assert!(x != y);
+        if x < y {
+            self.decrement((y, x), amount)
+        } else {
+            let i = self.index((x, y)) as usize;
+            self.cells[i] = self.cells[i].increment(amount)
+        }
+    }
+
+    pub fn decrement(&mut self, (x, y): (u32, u32), amount: Uint128) {
```

Comment:
since decrement could also increment, and increment could also decrement, could there be a more aptly named top-level function?

Reply:
i can't really think of one. they both do subtly different thing, and i'm pretty sure they just have to be mutually recursive like this. happy to add one though if you have a suggestion?

Reply:
IMO, it's fine for now, may just add a quick comment about it?